src/app/storage/PascalString.h

```diff
@@ -0,0 +1,139 @@
+/**
+ *    Copyright (c) 2025 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+#pragma once
+
+#include <algorithm>
+
+#include <app/ConcreteAttributePath.h>
+#include <cstdint>
+#include <lib/core/CHIPError.h>
+#include <lib/support/Span.h>
+
+namespace chip {
+namespace app {
+namespace Storage {
+
+/// Interprets a byte buffer as a pascal string with a 1-byte prefix.
+template <typename T>
+class ShortPascalString
+{
+public:
+    ShortPascalString(ShortPascalString &&)      = default;
+    ShortPascalString(const ShortPascalString &) = default;
+
+    template <size_t N>
+    ShortPascalString(T (&data)[N]) : mData(data), mMaxSize(N - 1)
+    {
+        static_assert(N > 0);
+        static_assert(N <= (1 + 0xFF)); // N-1 must fit in uint8_t
+    }
+
+    /// Returns the content of the pascal string.
+    /// Uses the prefix size information
+    Span<T> Content() { return { mData + 1, GetLength() }; }
+    Span<const T> Content() const { return { mData + 1, GetLength() }; }
+
+    /// Accesses the "PASCAL" string (i.e. valid data includint the string perfix)
+    Span<const T> PascalContent() const { return { mData, static_cast<size_t>(GetLength() + 1) }; }
+
+    /// Access to the full buffer. does NOT take into account current size
+    /// and includes the "size prefix"
+    Span<T> Buffer() { return { mData, static_cast<size_t>(mMaxSize + 1) }; }
+
+    uint8_t GetLength() const { return std::min(mMaxSize, (reinterpret_cast<const uint8_t *>(mData))[0]); }
+    bool SetLength(size_t len)
+    {
+        VerifyOrReturnError(len <= mMaxSize, false);
+        mData[0] = static_cast<T>(len);
+        return true;
+    }
+
+    bool SetValue(Span<const T> value)
+    {
+        if (!SetLength(value.size()))
+        {
+            return false;
+        }
+        memcpy(mData + 1, value.data(), value.size());
+        return true;
+    }
+
+private:
+    T * mData;
+    const uint8_t mMaxSize;
+};
+
+/// Interprets a byte buffer as a pascal string with a 2-byte prefix.
```

Comment:
So the only difference between ShortPascalString and LongPascalString is the prefix length constant (1 vs 2) and the 0xFF vs 0xFFFF in the static assert (which is `(1 < PrefixLen) - 1`).  Can we have:

```
namespace detail {
template <typename T, size_t PrefixLen>
class PascalString {
};
}
```

that holds most of the impl, instead of having two copies of everything here?  And then maybe 

````
template <typename T>
using LongPascalString = detail::PascalString<T, 2>;
````

and so on.  I guess SetLength's use of uint8_t vs uint16_t might need some finagling, as well as the return type of GetLength.  But with std::conditional_t defining a LengthType on the detail thing it should be doable, right?

Reply:
I started using size_t for all lengths though then, unsure if that is sane, will see if that can be changed.